lib/src/pubspec.dart

```diff
@@ -29,13 +30,53 @@ final _packageName = new RegExp(
 /// The default SDK upper bound constraint for packages that don't declare one.
 ///
 /// This provides a sane default for packages that don't have an upper bound.
-final _defaultUpperBoundSdkConstraint = new VersionConstraint.parse("<2.0.0");
+final VersionRange _defaultUpperBoundSdkConstraint =
+    new VersionConstraint.parse("<2.0.0");
+
+/// The upper bound contraint that matches the dev sdk.
+final _preReleaseTwoDotZeroSdkVersion = new Version.parse("2.0.0-dev.infinity");
+
+/// Whether or not to allow the pre-release sdk for packages that have an
+/// upper bound Dart SDK constraint of <2.0.0.
+///
+/// If enabled then a Dart SDK upper bound of <2.0.0 is always converted to
+/// <2.0.0-dev.infinity.
+///
+/// This has a default value of `true` but can be overridden with the
+/// PUB_ALLOW_PRERELEASE_SDK system environment variable.
+final bool allowPreReleaseTwoDotZeroSdk = () {
+  var userSetting =
+      Platform.environment["PUB_ALLOW_PRERELEASE_SDK"]?.toLowerCase() ?? 'true';
+  switch (userSetting) {
+    case "true":
+      return true;
+      break;
+    case "quiet":
+      warnAboutPreReleaseTwoDotZeroSdkOverrides = false;
+      return true;
+      break;
+    case "false":
+      return false;
+      break;
+    default:
+      warning(yellow('''
+The environment variable PUB_ALLOW_PRERELEASE_SDK is set as `$userSetting`.
+The expected value is either `true`, `quiet` (true but no logging), or `false`.
+Using a default value of `true`.
+'''));
+      return true;
+  }
+}();
+
+/// Whether or not to warn about pre-release sdk overrides.
+bool warnAboutPreReleaseTwoDotZeroSdkOverrides = true;
```

Comment:
This returns the wrong value if it's read before `allowPreReleaseTwoDotZeroSdk` is. Like @kevmoo suggests, you probably want a single lazy-initialized backing field and then two bool getters.

Reply:
done